Apply default semantic tags from linked channels to items

andrewfg opened the pull request:
Resolves #4898 

When an Item is linked to a Channel, and the Item does not have its own Point and/or Property tags, and the Item Channel Link is configured with `useTags=true`, then the default Point and/or Property tags of the Channel are applied to the Item.

This enhancement simplifies configuration for users by reducing the need for manual tag assignment.

The `useTags=true` setting can be applied individually for each channel link of a given Item. Or alternatively `useTags=true` can be applied system-wide globally for all Items via a setting in `runtime.cfg`

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

andrewfg commented:
@jimtng my edit to my prior comment crossed over with your subsequent comment, so please look up.

jimtng commented:
I would expect that the tags are combined with the custom tags that have been specified. But if they deleted the tags, especially in managed items (UI), then I think it should keep coming back on again until they remove the `useTags`.

In any case, I don't think it's currently possible to add custom config in the UI, so unless this is added, this is a moot issue?

jimtng commented:
To clarify, 

```
String MyItem [MyCustomTag] { channel="x:y:z" [ useTags=true ] }
```

Would result in MyItem having MyCustomTag + whatever the tags provided by the channel.

andrewfg commented:
> Would result in MyItem having MyCustomTag + whatever the tags provided by the channel.

That is not how I coded it currently; it is currently either/or. However I guess that the union of the two tag sets is as you say really the proper solution. (However we need to think of the case where there are two channels links both with useTags..)

andrewfg commented:
> the item registry should not be updated, it contains the tags set by users.

The tags can only be updated in the item registry if the user has set 'useTags' on the item channel link configuration.

andrewfg commented:
Aargh! I ran spotless and it chas changed 142 pom.xml files !!

andrewfg commented on itests/org.openhab.core.thing.tests/src/main/java/org/openhab/core/thing/link/ItemChannelLinkAssignDefaultTagsOSGiTest.java:
```diff
@@ -0,0 +1,201 @@
+        assertEquals(2, item.getTags().size());
+        assertTrue(item.getTags().contains("tag1"));
+        assertTrue(item.getTags().contains("tag2"));
+        assertTrue(item.getTags().contains("this should fail"));
```
@lolodomo / @jimtng as you can see this test does NOT fail .. which is a proof the the build system is not actually running this test suite. Indeed by placing a similar failing test in other OSGi test classes in `org.openhab.core.thing.tests` we can easily see that **_none of the OSGI tests in the module are actually being run._**

andrewfg replied:
https://github.com/openhab/openhab-core/issues/4921

lolodomo commented:
> I guess that you mean that if an item channel link would be removed then you want to revert to whatever were the prior tags?

Yes

> We need to test the timing too. Specifically to confirm that the item is in the registry before the link is created..

I have the same thought yesterday.
I hope it is the case.
To be checked for all item providers.

lolodomo commented:
And please rename the title of your PR as your change is not specific to .items file but to all config files defining items.

jimtng commented:
> > Alternatively I suppose we can assume that (in the case of .items files) if a link has been removed, it explicitly means that the .items file entry has been entirely reloaded, so there might be no need for local caching in the link registry..
> 
> Yes, makes sense for DSL and YAML.

We can't rely on this. Links can be added and removed on the fly, onto existing items, not just when the items were first created.



> Specifically to confirm that the item is in the registry before the link is created..

I'm not sure I understand what you're saying or what you intended to do, but links can exist even though items don't exist. We can have orphaned links. One example would be the links loaded from jsondb. This is why there's "Health" reporting in mainui.

andrewfg commented:
@lolodomo I have added support for link removal.

> not sure I understand what you're saying or what you intended to do, but links can exist even though items don't exist.

@jimtng when a link is created (or updated) we scan the linked channel for possible tags, and then we update the item in the item registry with those tags. So this tagging can only work if there is an actual item in the item registry. Therefore when a new item is created from (say) a .items file the sequence of object creation becomes important:

1. Item object created and added to item registry, then link object(s) created and added to link registry. => Works!
2. Link object(s) created and added to link registry, then item object created and added to item registry. => Fails!

andrewfg commented:
> when a new item is created from (say) a .items file the sequence of object creation becomes important

In the meantime I checked GenericItemProvider and YamlItemProvider .. and at least those two are first adding the item to the item registry and then adding the links the the link registry. So at least in those two cases the timing should be Ok.

Unfortunately OH core is too dense for me to figure out the timing for managed items. I horribly suspect that the items and the links are loaded separately and asynchronously (perhaps even in multiple threads). So I could imagine that we might have the cases of links being created before items here. I will think about possible methods to handle timing crossovers tomorrow. But in the meantime the good news is that managed items are probably not going to use such tag importation anyway, since the tags come via the UI.

... however it would be nice if either 1) we could build OH Core to test on a live system, and/or 2) we could run the OSGi tests .. hopefully tomorrow ??

andrewfg commented:
> this is getting more and more complex, is it worth it for the feature we're trying to add?

It was complex. But the code is written now.
 
> Is this something you or @lolodomo will use?

Sure.

andrewfg commented:
@lolodomo / @jimtng status update..

- code complete
- CI build running again
- still no OSGi tests running => hopefully someone will resolve #4921 as it is rather important
- tomorrow I will set up a new OH 5.1 test system and try to do the tests the hard way

jimtng commented on bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/link/ItemChannelLinkRegistry.java:
```diff
@@ -229,4 +252,139 @@ public enum ItemChannelLinkProblem {
+    }
+
+    /**
+     * If the linked channel has 'useTags=true' in its configuration return it's default tags.
```
```suggestion
     * If the linked channel has 'useTags=true' in its configuration return its default tags.
```

jimtng reviewed (commented):
I haven't looked into this closely, just a quick question:

This is a hypothetical scenario (haven't thought of exactly how to do this)
an Item is manually tagged with "CustomTag, Status, Power", then linked to a channel with useTags=true, and the channel provides "Status, Power" (matching what's set manually in the item).

Then what would happen when the link is removed. I would expect the item continues to have all 3 tags as manually set?

andrewfg commented:
> This is a hypothetical scenario

Your hypothesis is absolutely correct. This is indeed the one edge case where the deletion of a useTags channel link would indeed wrongly eliminate the native tags. This edge case could eventually be eliminated by adding a field to the item (say) ‘nativeTags’ but it would have to propagate through all aspects of items .. and this is probably indeed the limit point beyond which the extra development effort would exceed the extra benefit.

jimtng commented:
> his edge case could eventually be eliminated by adding a field to the item (say) ‘nativeTags’

How about in the itemchannellink itself, save the "original tags" before it added the channel's tags, and restore this on removal? This way you are keeping this data/logic contained within the same object.

andrewfg commented:
Yeah. I thought of that just after making my prior post. We could store it in the Configuration of the item channel link

jimtng commented:
Making it the default (with the caveat that it won't overwrite existing point/property) is convenient. It can also be a "surprise" although I think it should be a good kid of surprise.



> Yup. See my latest commit.

How does it handle the following scenarios:

Scenario A: Item defined with "Point", and "Property", linked to a channel with "Point" and "Property", then channel is unlinked -> Item should retain its "Point" and "Property" tags.

Scenario B: Item defined with "Point" only, linked to a channel with "Point" and "Property". When channel is unlinked -> Item should retain its "Point", whereas "Property" is removed.

andrewfg commented:
@jimtng I don't agree with either of your scenarios. If the item has native tags then they are never changed. But if the item has no native tags, then the channel tags are applied/removed all or nothing as the channel link is applied/removed.

jimtng commented:
> Making it the default (with the caveat that it won't overwrite existing point/property) is convenient. It can also be a "surprise" although I think it should be a good kind of surprise. But I'm not really sure of this. It would certainly cause an unexpected behaviour. So the prudent way is to make it opt-in.

@rkoshak, wdyt regarding whether we make this automatic tagging from the channel's default tags be the default (opt-out), or opt-in?

andrewfg commented:
> make people from both camps to be happy!

Are you sure that it is not too metaphysical? It would be simpler to have a fixed behavior..

andrewfg commented:
Ok. I did my first test on a live system. And I can confirm that the Item does take over the both the native tags on the item (green color) and the tags of the linked channel (red color) as shown below. Note: I am not sure why both sets of tags are listed in the UI under the title "Non-Semantic Tags" (blue color) .. perhaps that is another issue to be opened?

```
Color Office_Lamp_Colour "Office Lamp Colour" [foo, bar] {channel="hue:device:001788fffe2157c7:24e7a04e-ad36-4192-961e-ca7f22103cbc:color" [ useTags=true ] }
```

<img width="885" height="406" alt="image" src="https://github.com/user-attachments/assets/d036853b-6980-44ca-8a13-2fe991568dd3" />

lolodomo commented:
Maybe you added the tags and not replaced them?

lolodomo commented:
foo+bar should not be changed if the user assigned them to the item.

andrewfg commented:
> is there any way to do an info level log when the suggested semantic tags are applied

@rkoshak sure. It is currently logged at debug level but I will change that.

mherwege commented:
What happens if you have the same item linked to 2 channels with conflicting default tags?

andrewfg commented:
> That would make people from both camps to be happy!
> It could go in the runtime.cfg - since we're dealing with text-file people, it shouldn't be a problem.

@jimtng I don't know which OH core module is responsible for parsing such configs and making the available to other modules. => Can you give me a clue please?

jimtng commented:
> alwaysUseTags

I'm just thinking about making this as self-explanatory as possible. Alternatives are: `applyChannelTags`, `applyChannelTagsToItem`, `addChannelTags`, and`addChannelTagsToItem`

andrewfg commented:
My logic is that useTags applies with a certain syntax for individual links and that alwaysUseTags applies _with the exact same syntax_ for all links in the registry. If we rename the one, we should rename them both.

andrewfg commented on bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/link/ItemChannelLinkRegistry.java:
```diff
@@ -49,18 +62,44 @@
 @NonNullByDefault
-@Component(immediate = true, service = ItemChannelLinkRegistry.class)
-public class ItemChannelLinkRegistry extends AbstractLinkRegistry<ItemChannelLink, ItemChannelLinkProvider> {
+@Component(immediate = true, service = ItemChannelLinkRegistry.class, configurationPid = "org.openhab.core.things.ItemChannelLinkRegistry")
```
Ok

lolodomo commented on bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/link/ItemChannelLinkRegistry.java:
```diff
@@ -229,4 +270,171 @@ public enum ItemChannelLinkProblem {
+        Set<String> channelDefaultTags = getChannelDefaultTags(link);
+        if (!channelDefaultTags.isEmpty()) {
+            if (alreadyHasPointOrPropertyTag) {
+                logger.warn("Item '{}' already tagged; forbidden to assign tags from channel '{}'.",
+                        activeItem.getName(), link.getLinkedUID());
```
Why a warning?!
This is just a normal case, certainly even the standard case until this new feature is added.
This would generate massive logs for no reason as there is nothing to warn.
In my opinion, there is nothing to log, or just a DEBUG log.

andrewfg replied:
It means that the user has in the .items file the user self intentionally applied **both** Point/Property at the native item level and also declared useTags at the Channel level. So we have to warn the user that they cannot have both.

HOWEVER having said that, if we go the way of applying alwaysUseTags at global level then indeed we should probably not log it.

lolodomo replied:
> HOWEVER having said that, if we go the way of applying alwaysUseTags at global level then indeed we should probably not log it.

Yes
I assume that most of users will enable this setting if they are informed of its existence, at least of course I will myself.

andrewfg replied:
Moi aussi

lolodomo commented on bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/link/ItemChannelLinkRegistry.java:
```diff
@@ -229,4 +270,171 @@ public enum ItemChannelLinkProblem {
+                    Set<String> otherLinkTags = getChannelDefaultTags(otherLink);
+                    if (!otherLinkTags.isEmpty()) {
+                        if (alreadyHasPointOrPropertyTag) {
+                            logger.warn("Item '{}' already tagged; forbidden to assign tags from channel '{}'.",
+                                    activeItem.getName(), otherLink.getLinkedUID());
```
Same here, no need for a warning

andrewfg replied:
Same comment as above..

andrewfg replied:
Is it now OK? With the conditional check?

lolodomo commented on bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/link/ItemChannelLinkRegistry.java:
```diff
@@ -229,4 +270,171 @@ public enum ItemChannelLinkProblem {
+                    oldLink.getLinkedUID());
+
+            // iterate over other links in case one may assign new tags
+            boolean alreadyHasPointOrPropertyTag = false;
```
Here,n you are not checking if a Point/Property tag is defined on the item itself like you did in `assignChannelDefaultTags`. There is something missing I believe.

andrewfg commented on bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/link/ItemChannelLinkRegistry.java:
```diff
@@ -229,4 +270,171 @@ public enum ItemChannelLinkProblem {
+            Set<String> newTags = new HashSet<>(activeItem.getTags());
+
+            // remove old link's tags
+            Set<String> oldLinkTags = getChannelDefaultTags(oldLink);
```
Good catch. Also in that case we don't need to iterate over other channels, since if they would have been providing the tags they can continue to do so.

andrewfg commented:
> My pref: applyChannelTagsToItem

Ok. But then we should use the same `applyChannelTagsToItem` for both the config at individual link level and also the config at global registry level. (My reason for having a shorter key name was to avoid too much typing for the user..)

andrewfg commented:
So again I suggest to use useTags for both the individual links scope and the global all links scope.

lolodomo commented:
Code looks good to me.

andrewfg commented:
Many thanks @lolodomo 

Now all I need to do is fix the iTests not running problem in #4921 (and of course pass them) .. and in light of the knowledge gained from the discussion in #4928 I will try to fix the test class to use the proposed hack.

I will let you know when it is ready..

andrewfg commented:
Ok. We are nearly there. Below are the test cases that I am using. I had to make a few more changes, but it does now pass all the tests.

<details><summary>Test case .things file</summary>

```
// in the definitions below you need to set the applicationKey and resourceIds to match your actual hardware
Bridge hue:bridge-api2:bridge "Bridge" [ipAddress="123.123.123.123", applicationKey="REDACTED"] {
    Thing device light "Light" [resourceId="REDACTED"]
    Thing device dimmer "Dimmer" [resourceId="REDACTED"]
}
```

</details>

<details><summary>Test case .items file</summary>

```
String Item_01 "Expected: String" {channel="hue:device:bridge:light:color" }

String Item_02 "Expected: String" {channel="hue:device:bridge:light:color" [useTags=false] }
String Item_03 "Expected: String" {channel="hue:device:bridge:light:color" [useTags="false"] }

String Item_04 "Expected: String · Point > Control > Color" {channel="hue:device:bridge:light:color" [useTags=true] }
String Item_05 "Expected: String · Point > Control > Color" {channel="hue:device:bridge:light:color" [useTags="true"] }

String Item_06 "Expected: String · Point > Control > Color; Custom" ["Custom"] {channel="hue:device:bridge:light:color" [useTags=true] }

// property without a point: OH magically provides 'Control' property
String Item_07 "Expected: String · Point > 'Control' > Power; Custom" ["Power", "Custom"] {channel="hue:device:bridge:light:color" }
String Item_08 "Expected: String · Point > 'Control' > Power; Custom" ["Power", "Custom"] {channel="hue:device:bridge:light:color" [useTags=true] }

String Item_09 "Expected: String · Point > Switch; Custom" ["Switch", "Custom"] {channel="hue:device:bridge:light:color" }
String Item_10 "Expected: String · Point > Switch; Custom" ["Switch", "Custom"] {channel="hue:device:bridge:light:color" [useTags=true] }

String Item_11 "Expected: String · Point > Switch > Power; Custom" ["Switch", "Power", "Custom"] {channel="hue:device:bridge:light:color" }
String Item_12 "Expected: String · Point > Switch > Power; Custom" ["Switch", "Power", "Custom"] {channel="hue:device:bridge:light:color" [useTags=true] }

String Item_13 "Expected: String · Point > Alarm > LowBattery" {channel="hue:device:bridge:dimmer:battery-low" [useTags=true] }

String Item_14 "Expected: String · Point > Control > Color" {channel="hue:device:bridge:light:color" [useTags=true], channel="hue:device:bridge:dimmer:battery-low" [useTags=true] }
String Item_15 "Expected: String · Point > Alarm > LowBattery" {channel="hue:device:bridge:light:color" [useTags=false], channel="hue:device:bridge:dimmer:battery-low" [useTags=true] }
```

</details>

<details><summary>Logger output</summary>

```
2025-07-26 16:26:51.460 [DEBUG] [e.thing.link.ItemChannelLinkRegistry] - Item 'Item_12' already tagged; ignoring tags supplied by channel 'hue:device:bridge:light:color'.
2025-07-26 16:26:51.462 [DEBUG] [e.thing.link.ItemChannelLinkRegistry] - Item 'Item_08' already tagged; ignoring tags supplied by channel 'hue:device:bridge:light:color'.
2025-07-26 16:26:51.463 [INFO ] [e.thing.link.ItemChannelLinkRegistry] - Item 'Item_05' adding tags '[Control, Color]' supplied by channel 'hue:device:bridge:light:color'.
2025-07-26 16:26:51.469 [DEBUG] [e.thing.link.ItemChannelLinkRegistry] - Item 'Item_10' already tagged; ignoring tags supplied by channel 'hue:device:bridge:light:color'.
2025-07-26 16:26:51.469 [INFO ] [e.thing.link.ItemChannelLinkRegistry] - Item 'Item_04' adding tags '[Control, Color]' supplied by channel 'hue:device:bridge:light:color'.
2025-07-26 16:26:51.472 [INFO ] [e.thing.link.ItemChannelLinkRegistry] - Item 'Item_14' adding tags '[LowBattery, Alarm]' supplied by channel 'hue:device:bridge:dimmer:battery-low'.
2025-07-26 16:26:51.475 [DEBUG] [e.thing.link.ItemChannelLinkRegistry] - Item 'Item_14' already tagged; ignoring tags supplied by channel 'hue:device:bridge:light:color'.
2025-07-26 16:26:51.475 [INFO ] [e.thing.link.ItemChannelLinkRegistry] - Item 'Item_06' adding tags '[Control, Color]' supplied by channel 'hue:device:bridge:light:color'.
2025-07-26 16:26:51.477 [INFO ] [e.thing.link.ItemChannelLinkRegistry] - Item 'Item_15' adding tags '[LowBattery, Alarm]' supplied by channel 'hue:device:bridge:dimmer:battery-low'.
2025-07-26 16:26:51.480 [INFO ] [e.thing.link.ItemChannelLinkRegistry] - Item 'Item_13' adding tags '[LowBattery, Alarm]' supplied by channel 'hue:device:bridge:dimmer:battery-low'.
```

</details>

andrewfg commented:
See https://github.com/openhab/openhab-docs/pull/2529

andrewfg commented:
I am not sure why the CI build fails with an error on the voice integration tests. I did nothing that should impact that part. ??

andrewfg commented:
I also finally figured out how to add the OSGi integration test (and it passes).

andrewfg commented:
This pull request introduces significant enhancements to the `ItemChannelLinkRegistry` and related classes to support automatic assignment and management of default semantic tags for items based on their linked channels. Additionally, it includes minor dependency updates in test configurations. Below is a summary of the most important changes:

### Enhancements to `ItemChannelLinkRegistry` and Tag Management:

1. **Automatic Tag Assignment and Management**:
   * Introduced methods to assign default semantic tags from linked channels to items, provided the channel has `useTags=true` and the item does not already have conflicting tags. Warnings are logged for conflicting tags or multiple linked channels with `useTags=true`. (`[bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/link/ItemChannelLinkRegistry.javaR285-R487](diffhunk://#diff-3c6274c126dc7186b400a8a006e8f8849b086463034900a11730973419ec003fR285-R487)`)
   * Added logic to remove tags supplied by a channel when the link is removed or updated, ensuring native custom tags remain intact. (`[bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/link/ItemChannelLinkRegistry.javaR285-R487](diffhunk://#diff-3c6274c126dc7186b400a8a006e8f8849b086463034900a11730973419ec003fR285-R487)`)
   * Implemented global and per-channel configuration for tag usage via the `USE_TAGS` property. (`[[1]](diffhunk://#diff-3c6274c126dc7186b400a8a006e8f8849b086463034900a11730973419ec003fR70-R114)`, `[[2]](diffhunk://#diff-3c6274c126dc7186b400a8a006e8f8849b086463034900a11730973419ec003fR285-R487)`)

2. **Event Handling and Registry Integration**:
   * Registered the `ItemChannelLinkRegistry` as a listener for item registry changes and system start level events to manage tags dynamically during item lifecycle events (addition, removal, updates). (`[[1]](diffhunk://#diff-3c6274c126dc7186b400a8a006e8f8849b086463034900a11730973419ec003fR70-R114)`, `[[2]](diffhunk://#diff-3c6274c126dc7186b400a8a006e8f8849b086463034900a11730973419ec003fR285-R487)`)
   * Added `notifyListenersAboutItemExternalUpdate` to `ItemRegistry` for notifying listeners of external item updates, ensuring proper synchronization. (`[[1]](diffhunk://#diff-6b9dd17d5481dad9b22edc83b356d73477460b15c8b4f4bb30ba2fdd64983158R467-R471)`, `[[2]](diffhunk://#diff-856755163fef4f78f17c08b696bc172894935c6761a66c9a1205454dee2d043cR108-R114)`)

3. **Refactoring and Code Simplification**:
   * Simplified stream operations in `getLinkedItems` and `getBoundThings` methods for better readability. (`[[1]](diffhunk://#diff-3c6274c126dc7186b400a8a006e8f8849b086463034900a11730973419ec003fL89-R141)`, `[[2]](diffhunk://#diff-3c6274c126dc7186b400a8a006e8f8849b086463034900a11730973419ec003fL100-R152)`)
   * Added a transient `tagsLinked` flag to `ItemChannelLink` to track whether tags were assigned by a specific channel. (`[[1]](diffhunk://#diff-a985152020c0e792d8079ee081e17eae000b8d2e0a8bfbb1b927fe9252eecc2fR34-R35)`, `[[2]](diffhunk://#diff-a985152020c0e792d8079ee081e17eae000b8d2e0a8bfbb1b927fe9252eecc2fR77-R87)`)

### Dependency Updates in Test Configurations:

4. **Test Configuration Updates**:
   * Updated dependencies in `itest.bndrun` files to include `org.openhab.core.model.persistence.runtime` and remove unused `org.openhab.core.model.rule.runtime`. (`[[1]](diffhunk://#diff-0d15e70053365c51dfcffdb077af6e8c4c79c365cb4ddb2fe3e085033806f079L127-R132)`, `[[2]](diffhunk://#diff-e4b49351566f44aa46c31af7a2cee504af1b7441368afc2a4296b6049c1cb3bfL134-R135)`, `[[3]](diffhunk://#diff-9eaac1e2084455180924aa70459c1bf05c71197553c619d9bb75229564507186L133-R138)`, `[[4]](diffhunk://#diff-41e516a93b6d481f8b4babc9d49c8ebd022942000d046d65f73f8d769d672f34L136-R141)`)

lolodomo commented:
> I am guessing the OH core maintainers are looking for signs of qualified support from other OH experts. So @lolodomo it might help if you want this if you could give a formal LGTM review..

How can I identify all your commits after my previous "Code looks good to me" ?

andrewfg commented:
> How can I identify all your commits after my previous "Code looks good to me" ?

Umm. Probably with great difficulty. So I would suggest to look at the last final state of the 12 modified files.

lolodomo commented:
Why are you not pushing commits as we all do which means the timing is kept and we as reviewers immediately see what commits are not yet read ?

andrewfg commented:
> Why are you not pushing commits as we all do

Sorry but I don't understand the question. Can you please explain?

jimtng commented on itests/org.openhab.core.thing.tests/src/main/java/org/openhab/core/thing/link/ItemChannelLinkTaggingOSGiTest.java:
```diff
@@ -0,0 +1,176 @@
+        item = itemRegistry.get("Item_07");
+        assertNotNull(item);
+        assertTrue(item.getTags().contains("Power"));
+        assertTrue(item.getTags().contains("Custom"));
```
There was a piece of code in core that automatically adds a Point tag (Measurement, or Control) if an item only has Property tag.

jimtng replied:
I just checked, I think it's still there

lolodomo replied:
Where please ?

andrewfg replied:
It is obviously in the codepath after the code used by the itest. (But I suppose that could change; so I think it is better to revert these two tests just in case)..

lolodomo replied:
I don't understand. This is not code in core framework?

andrewfg replied:
The itest does indeed use (part of) the core model parser to create its test Items. However it apparently takes the Items from quite deep in the core code, presumably before any auto magic corrections are applied, and before any validation checks and failures would be applied.

The purpose of these tests is NOT to test the model parser but to test the useTags functionality. So we should structure the tests to ONLY check stuff that may or may not have been contributed via useTags.

jimtng replied:
I finally found it. It doesn't actually add the tag, but returning an item as a point even when a Point tag is missing, with only a property tag

https://github.com/openhab/openhab-core/blob/8537a1b82278159d7ec69c1bb45508ca2c9537a1/bundles/org.openhab.core.semantics/src/main/java/org/openhab/core/semantics/SemanticTags.java#L64-L82

So I guess that means it won't affect you and you could do `assertEquals(Set.of("Power", "Custom"), item.getTags())` like @lolodomo mentioned if you prefer. But I'm not making a suggestion one way or another.

andrewfg commented:
TODO in cases where the user has configured useTags=true at a global level, then it shall be possible to set useTags=false at item level in order to configure an item specific exception to the global rule.

lolodomo commented:
> TODO in cases where the user has configured useTags=true at a global level, then it shall be possible to set useTags=false at item level in order to configure an item specific exception to the global rule.

Ah yes, if not yet possible, I missed something very important..

andrewfg commented on itests/org.openhab.core.model.script.tests/itest.bndrun:
```diff
@@ -135,4 +134,11 @@ Fragment-Host: org.openhab.core.model.script
 	com.fasterxml.jackson.dataformat.jackson-dataformat-yaml;version='[2.19.2,2.19.3)',\
-	org.yaml.snakeyaml;version='[2.4.0,2.4.1)'
+	org.yaml.snakeyaml;version='[2.4.0,2.4.1)',\
+	org.openhab.core.model.persistence.runtime;version='[5.1.0,5.1.1)'
```
Done

andrewfg commented on bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/link/ItemChannelLinkRegistry.java:
```diff
@@ -229,4 +282,208 @@ public enum ItemChannelLinkProblem {
+     * the log. The warning is also logged if the item has more than one linked channel with 'useTags=true'.
+     * And in any case if the item has native custom tags then those tags remain.
+     */
+    private void assignChannelDefaultTags(ItemChannelLink link, ActiveItem activeItem) {
+        boolean alreadyHasPointOrPropertyTag = false;
+
+        for (String tag : activeItem.getTags()) {
+            Class<? extends Tag> type = SemanticTags.getById(tag);
+            if ((type != null) && (Point.class.isAssignableFrom(type) || Property.class.isAssignableFrom(type))) {
```
Caching is not a good idea. Suggestion rejected.

holgerfriedrich commented:
> @wborn it seems that @holgerfriedrich is pushing dependency updates faster than I can follow with fixing the bndrun files ;) so I can't get it to run the iTests => let us wait until tomorrow..

@andrewfg sorry for that! It seems today is patchday, bringing you all the latest features (and bugs) of the week ;-) ;-)
We are almost finished with the planned upgrades for 5.1, so it should be more relaxed in a few days....

andrewfg commented:
@holgerfriedrich / @wborn many thanks for the support; it is no all fixed and builds without error..

holgerfriedrich commented:
@andrewfg There are again some conflicts in the itests, caused by me in #5005.... Sorry for that.

andrewfg commented:
> not available to fix the conflicts soon ? The milestone build will start very soon

@lolodomo I finally just did it

andrewfg commented:
@holgerfriedrich can you please do a final check that my fix to your break of my pr did not break yours..

andrewfg commented:
@lolodomo / @holgerfriedrich  / @kaikreuzer => Can you please not leave it until the last minute before a milestone before changing the itests thus breaking the build and preventing this PR from ever being merged? Can you please either merge it now into a snapshot, or at least give me 24 hours to sort out last minute merge conflicts?

lolodomo commented:
@openhab/core-maintainers : can you merge it please? I would like to test it in a snapshot before the next milestone.
